### PR TITLE
Fix: prevent space-before-function-paren from checking type annotations

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] Bug fix

**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** 7.7.4
* **npm Version:** 4.1.2

**What parser (default, Babel-ESLint, etc.) are you using?**

babel-eslint

**Please show your full configuration:**

```yml
parser: babel-eslint
rules:
  space-before-function-paren: error
```

**What did you do? Please include the actual source code causing the issue.**

```js
type TransformFunction = (el: ASTElement, code: string) => string;
```

**What did you expect to happen?**

I expected ESLint to not crash.

**What actually happened? Please include the actual, raw output from ESLint.**

ESLint crashed.

```
TypeError: Cannot read property 'range' of null
    at TokenStore.getTokenBefore (/path/to/eslint-canary/.downloaded-projects/vue/node_modules/eslint/lib/token-store/index.js:318:17)
    at EventEmitter.:function (/path/to/eslint-canary/.downloaded-projects/vue/node_modules/eslint/lib/rules/space-before-function-paren.js:116:46)
    at emitOne (events.js:96:13)
    at EventEmitter.emit (events.js:191:7)
    at NodeEventGenerator.applySelector (/path/to/eslint-canary/.downloaded-projects/vue/node_modules/eslint/lib/util/node-event-generator.js:265:26)
    at NodeEventGenerator.applySelectors (/path/to/eslint-canary/.downloaded-projects/vue/node_modules/eslint/lib/util/node-event-generator.js:292:22)
    at NodeEventGenerator.enterNode (/path/to/eslint-canary/.downloaded-projects/vue/node_modules/eslint/lib/util/node-event-generator.js:308:14)
    at CodePathAnalyzer.enterNode (/path/to/eslint-canary/.downloaded-projects/vue/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:602:23)
    at CommentEventGenerator.enterNode (/path/to/eslint-canary/.downloaded-projects/vue/node_modules/eslint/lib/util/comment-event-generator.js:98:23)
    at Traverser.enter (/path/to/eslint-canary/.downloaded-projects/vue/node_modules/eslint/lib/eslint.js:929:36)
```

This was working in 3.18.0, but it broke due to 9b509ce27945a8c5a3d0516d0ea134950897f21c.

**What changes did you make? (Give an overview)**

Previously, the space-before-function-paren rule's listener would be triggered for custom nodes such as `FunctionTypeAnnotation`, and would throw an error. This broke the regression build. This commit updates the rule to only listen for known function nodes.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular

### CODE REVIEW

**Comment by ilyavolodin** [comment]
I think at some point we need to figure out a better way to test those changes. Those fixtures are huge.

**Comment by not-an-aardvark** [comment]
We could add babel-eslint as a devDependency for testing and have it handle the parsing on its own.

File: tests/fixtures/parsers/space-before-function-paren/function-type-annotation.js [hunk 1]
```diff
@@ -0,0 +1,525 @@
+"use strict";
```
Comment (vitorbal) [inline tests/fixtures/parsers/space-before-function-paren/function-type-annotation.js]: I'm not sure, but would it make sense to name the folder `tests/fixtures/parsers/babel/` instead? Or maybe we could clean up/re-organize the `fixtures/parsers/` folder in a separate PR. I think it's a bit confusing now where to put parser fixtures.. related: https://github.com/eslint/eslint/pull/8297/files#diff-824f19921586246afb6f381021c34222

Comment (not-an-aardvark) [inline tests/fixtures/parsers/space-before-function-paren/function-type-annotation.js]: Good idea, I moved the fixture. I agree with reorganizing the folder in a separate PR.

**Pull request merged**